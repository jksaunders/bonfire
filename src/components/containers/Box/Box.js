import React from 'react';
import PropTypes from 'prop-types';
import styled from 'styled-components';
import { css, cssBackground, cssBorder, CssRuleType } from '../../../styling';

export const BoxPropTypes = {
  align: CssRuleType(PropTypes.string),
  background: CssRuleType(PropTypes.string),
  border: CssRuleType(PropTypes.string),
  borderRadius: CssRuleType(PropTypes.string),
  children: PropTypes.node,
  className: CssRuleType(PropTypes.string),
  full: PropTypes.bool,
  fullHeight: PropTypes.bool,
  fullWidth: PropTypes.bool,
  height: CssRuleType(PropTypes.string),
  margin: CssRuleType(PropTypes.string),
  maxHeight: CssRuleType(PropTypes.string),
  maxWidth: CssRuleType(PropTypes.string),
  minHeight: CssRuleType(PropTypes.string),
  minWidth: CssRuleType(PropTypes.string),
  overflow: CssRuleType(PropTypes.string),
  overflowX: CssRuleType(PropTypes.string),
  overflowY: CssRuleType(PropTypes.string),
  padding: CssRuleType(PropTypes.string),
  useTypography: PropTypes.bool,
  width: CssRuleType(PropTypes.string),
};

export const BoxDefaultProps = Object.fromEntries(
  Object.entries(BoxPropTypes).map(e => [e[0], null])
);

const cssReset = `
  margin: 0;
  padding: 0;
  border: 0;
  font-size: 100%;
  font: inherit;
  vertical-align: baseline;
`;

const StyledBox = styled.div`
  ${/* reset */ ''}
  ${cssReset}
  
  box-sizing: border-box;
  ${cssBackground}
  ${cssBorder}
  ${css('borderRadius', 'border-radius')}

  ${/* sizing */ ''}
  ${css(
    ['full', 'height', '100%'],
    ['fullHeight', 'height', '100%'],
    ['height']
  )}
  ${css(['full', 'width', '100%'], ['fullWidth', 'width', '100%'], ['width'])}
  ${css('maxHeight', 'max-height')}
  ${css('maxWidth', 'max-width')}
  ${css('minHeight', 'min-height')}
  ${css('minWidth', 'min-width')}
  ${css('margin')}
  ${css('padding')}

  ${/* overflow */ ''}
  ${css('overflow')}
  ${css('overflowX', 'overflow-x')}
  ${css('overflowY', 'overflow-y')}
`;

const Box = React.forwardRef(({ children, component, ...rest }, ref) => (
  <StyledBox as={component} ref={ref} {...rest}>
    {children}
  </StyledBox>
));

Box.propTypes = { ...BoxPropTypes, component: PropTypes.string.isRequired };
Box.defaultProps = BoxDefaultProps;

export default Box;
