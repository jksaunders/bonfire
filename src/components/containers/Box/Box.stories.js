import React from 'react';
import styled from 'styled-components';

import { boolean, number, text, withKnobs } from '@storybook/addon-knobs';

import Box from './Box';
import Typography, { MaterialVariants } from '../../Typography';

const Content = styled.div`
  background-color: rgba(255, 255, 255, 0.7);
`;

export default {
  title: 'Containers|Box',
  component: Box,
  decorators: [withKnobs],
};

export const BasicStory = () => {
  const elevation = number('Elevation', 0);
  const full = boolean('Full height and width?', false);
  const padding = text('Padding', '');

  return (
    <Box
      elevation={elevation}
      full={full}
      padding={padding !== '' ? padding : null}
    >
      <Content>
        <Typography variant={MaterialVariants.H2}>content1</Typography>
      </Content>
      <Content>
        <Typography variant={MaterialVariants.H2}>content2</Typography>
      </Content>
      <Content>
        <Typography variant={MaterialVariants.H2}>content3</Typography>
      </Content>
    </Box>
  );
};

BasicStory.story = {
  name: 'Basic',
};
