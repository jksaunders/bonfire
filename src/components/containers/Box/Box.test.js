import React from 'react';
import { expectExists, expectSnapshot, render } from '../../../testing';
import Box from './Box';

const getBox = (props = {}, children) =>
  render(<Box {...props}>{children}</Box>);

test('renders', () => {
  const { component } = getBox({ component: 'input' });
  expectExists(component);
  expectSnapshot(component);
});

describe('positioning', () => {
  test('position', () => {
    const { component } = getBox({ position: 'relative' });
    expectExists(component);
    expectSnapshot(component);
    expect(component).toHaveStyleRule('position', 'relative');
  });

  test('absolute', () => {
    const { component } = getBox({ absolute: true });
    expectExists(component);
    expectSnapshot(component);
    expect(component).toHaveStyleRule('position', 'absolute');
  });

  test('top 2', () => {
    const { component } = getBox({
      absolute: true,
      absoluteAnchor: true,
    });
    expectExists(component);
    expectSnapshot(component);
    expect(component).toHaveStyleRule('position', 'absolute');
  });

  test('all 3', () => {
    const { component } = getBox({
      absolute: true,
      absoluteAnchor: true,
      position: 'unset',
    });
    expectExists(component);
    expectSnapshot(component);
    expect(component).toHaveStyleRule('position', 'absolute');
  });
});
