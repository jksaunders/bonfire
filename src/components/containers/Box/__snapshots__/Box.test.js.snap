// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`positioning absolute 1`] = `
.c0 {
  box-sizing: border-box;
  position: absolute;
}

<div
  class="c0"
/>
`;

exports[`positioning all 3 1`] = `
.c0 {
  box-sizing: border-box;
  position: absolute;
}

<div
  class="c0"
/>
`;

exports[`positioning position 1`] = `
.c0 {
  box-sizing: border-box;
  position: relative;
}

<div
  class="c0"
/>
`;

exports[`positioning top 2 1`] = `
.c0 {
  box-sizing: border-box;
  position: absolute;
}

<div
  class="c0"
/>
`;

exports[`renders 1`] = `
.c0 {
  box-sizing: border-box;
}

<input
  class="c0"
/>
`;
